p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,242 @@
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ tpt.CapableConn = &connection{}
+
+const (
+	maxStreamBufferSize int = 10
+)
+
+type connection struct {
+	pc        *webrtc.PeerConnection
+	transport *WebRTCTransport
+	scope     network.ConnManagementScope
+
+	localPeer      peer.ID
+	privKey        ic.PrivKey
+	localMultiaddr ma.Multiaddr
+
+	remotePeer      peer.ID
+	remoteKey       ic.PubKey
+	remoteMultiaddr ma.Multiaddr
+
+	m       sync.Mutex
+	streams map[uint16]*dataChannel
+
+	streamChan chan network.MuxedStream
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newConnection(
+	pc *webrtc.PeerConnection,
+	transport *WebRTCTransport,
+	scope network.ConnManagementScope,
+
+	localPeer peer.ID,
+	privKey ic.PrivKey,
+	localMultiaddr ma.Multiaddr,
+
+	remotePeer peer.ID,
+	remoteKey ic.PubKey,
+	remoteMultiaddr ma.Multiaddr,
+) *connection {
+	streamChan := make(chan network.MuxedStream, maxStreamBufferSize)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	conn := &connection{
+		pc:        pc,
+		transport: transport,
+		scope:     scope,
+
+		localPeer:      localPeer,
+		privKey:        privKey,
+		localMultiaddr: localMultiaddr,
+
+		remotePeer:      remotePeer,
+		remoteKey:       remoteKey,
+		remoteMultiaddr: remoteMultiaddr,
+		ctx:             ctx,
+		cancel:          cancel,
+		streams:         make(map[uint16]*dataChannel),
+
+		streamChan: streamChan,
+	}
+
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		id := *dc.ID()
+		var stream *dataChannel
+		dc.OnOpen(func() {
+			// datachannel cannot be detached before opening
+			rwc, err := dc.Detach()
+			if err != nil {
+				log.Errorf("[%s] could not detch channel: %s", localPeer, dc.Label())
+				return
+			}
+			stream = newDataChannel(conn, dc, rwc, pc, nil, nil)
+			conn.addStream(id, stream)
+			streamChan <- stream
+		})
+
+	})
+
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		if state == webrtc.PeerConnectionStateClosed || state == webrtc.PeerConnectionStateDisconnected {
+			conn.Close()
+		}
+	})
+
+	return conn
+}
+
+// ConnState implements transport.CapableConn
+func (c *connection) ConnState() network.ConnectionState {
+	return network.ConnectionState{}
+}
+
+// Implement network.MuxedConn
+
+func (c *connection) Close() error {
+	if c.IsClosed() {
+		return nil
+	}
+
+	c.scope.Done()
+	c.cancel()
+	_ = c.pc.Close()
+	return nil
+}
+
+func (c *connection) IsClosed() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+	}
+	return false
+}
+
+func (c *connection) OpenStream(ctx context.Context) (network.MuxedStream, error) {
+	type openStreamResult struct {
+		network.MuxedStream
+		error
+	}
+
+	if c.IsClosed() {
+		return nil, os.ErrClosed
+	}
+
+	result := make(chan openStreamResult)
+	dc, err := c.pc.CreateDataChannel("", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	streamID := *dc.ID()
+	var stream *dataChannel
+	dc.OnOpen(func() {
```

Comment:
It's probably easier to _just_ close a channel here and handle the rest of the setup outside of this callback.

Reply:
We can't call `Detach` outside `OnOpen`. https://github.com/pion/webrtc/blob/master/datachannel.go#L192-L216. I need the detached channel for stream level congestion control.